### PR TITLE
Picker view for restricted textfields

### PR DESCRIPTION
**Pull request by stowellm**
Addresses #77 by replacing the keyboard with a UIPickerView selection utility that only allows users to choose between any of the textual restrictions placed on a field within a project.

On the side, also addresses #43 by adding a simple confirmation dialog to inform the user that if he/she has data still in textfields but is trying to navigate to upload instead of the 'save row' or 'save data set' buttons.

### CODE REVIEW

File: Apps/Writer/Writer/ISWViewController.m [hunk 1]
```diff
@@ -307,9 +307,34 @@ - (IBAction)credentialBarBtnOnClick:(id)sender {
 
 - (IBAction)uploadBtnOnClick:(id)sender {
 
-    QueueUploaderView *queueUploader = [[QueueUploaderView alloc] initWithParentName:PARENT_WRITER andDelegate:self];
-    queueUploader.title = @"Upload";
-    [self.navigationController pushViewController:queueUploader animated:YES];
+    // check if the user has any data left unsaved first
+    if (dataArr != nil && dataArr.count > 0) {
```
Comment (jaypoulz) [inline Apps/Writer/Writer/ISWViewController.m]: :cactus:

File: Apps/Writer/Writer/ISWViewController.m [hunk 2]
```diff
@@ -307,9 +307,34 @@ - (IBAction)credentialBarBtnOnClick:(id)sender {
...
+
+        for (FieldData *data in dataArr) {
+
+            if (data != nil) {
```
Comment (jaypoulz) [inline Apps/Writer/Writer/ISWViewController.m]: :cactus:

File: Apps/Writer/Writer/ISWViewController.m [hunk 3]
```diff
@@ -307,9 +307,34 @@ - (IBAction)credentialBarBtnOnClick:(id)sender {
...
+
+                int type = data.fieldType.intValue;
+                NSString *dataStr = data.fieldData;
+
+                if ((type == TYPE_NUMBER || type == TYPE_TEXT) && dataStr.length > 0) {
```
Comment (jaypoulz) [inline Apps/Writer/Writer/ISWViewController.m]: :cat:

File: Apps/Writer/Writer/ISWViewController.m [hunk 4]
```diff
@@ -542,9 +575,52 @@ - (UIToolbar *) createDoneKeyboardView {
 
 #pragma mark - UITextField code
 
+- (BOOL)textFieldShouldBeginEditing:(UITextField *)textField {
+
+    // check if the text field has any restrictions
+    if (textField.tag != kDATA_SET_NAME_TAG) {
```
Comment (jaypoulz) [inline Apps/Writer/Writer/ISWViewController.m]: :cactus:

File: Apps/Writer/Writer/ISWViewController.m [hunk 5]
```diff
@@ -542,9 +575,52 @@ - (UIToolbar *) createDoneKeyboardView {
...
+
+        FieldData *data = [dataArr objectAtIndex:textField.tag];
+        NSArray *res = data.fieldRestrictions;
+
+        if (res != (id)[NSNull null] && res != nil && res.count > 0) {
```
Comment (jaypoulz) [inline Apps/Writer/Writer/ISWViewController.m]: :bug:

File: Apps/Writer/Writer/ISWViewController.m [hunk 6]
```diff
@@ -813,6 +902,53 @@ - (UIAlertView *)getDispatchDialogWithMessage:(NSString *)dString {
...
+
+    return pickerDataSource[row];
+}
+
+// removes leading and trailing whitespace from elements in the data source array
+- (void) trimDataSourceWhiteSpace {
+
+    NSMutableArray *mutableDataSource = [pickerDataSource mutableCopy];
+
+    if (pickerDataSource != (id)[NSNull null] && pickerDataSource != nil) {
```
Comment (jaypoulz) [inline Apps/Writer/Writer/ISWViewController.m]: :cactus:

**Comment by stowellm** [comment]
@jaypoulz

**Pull request merged**